### ISSUE TITLE
Frequent `Killed: 9` with rust on M1 Mac

### ISSUE THREAD

**Issue by n8henrie** [NixOS/nixpkgs#162795]
### Describe the bug

I'm writing my first `shell.nix`:

```nix
{ pkgs ? import <nixpkgs> { } }:
with pkgs;
pkgs.mkShell {
  nativeBuildInputs = [
    cargo
    cargo-edit
    libiconv
    openssl
    pkg-config
    rust-analyzer
    rustc
    rustfmt
    darwin.apple_sdk.frameworks.Security
  ];
}
```

It runs a (currently trivial) Rust script:

```rust
use rss::Channel;

async fn get_channel() -> anyhow::Result<Channel> {
    let ua = "killed-9-example";
    let client = reqwest::Client::builder().user_agent(ua).build()?;
    let content = client
        .get("https://n8henrie.com/feed.xml")
        .send()
        .await?
        .bytes()
        .await?;
    let channel = Channel::read_from(&content[..])?;
    Ok(channel)
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let result = get_channel().await?;
    for item in result.items() {
        for cat in &item.categories {
            println!("{:?}", cat)
        }
    }
    Ok(())
}
```

The code runs and works as expected. Sometimes. Unfortunately, I find that when I make a few trivial changes, `cargo run` results in `Killed: 9`. `cargo clean && cargo run` and it usually works again.

I made an example repo <https://github.com/n8henrie/killed-9-example> -- though it looks like an even more trivial example reproduces the bug (example below).

### Steps To Reproduce

```console
$ git clone https://github.com/n8henrie/killed-9-example
$ cd ./killed-9-example
$ nix-shell
$ cargo run
... it works ... 
$ nvim -u NONE src/main.rs 
... make a trivial change ... 
$ cargo run
   Compiling killed-9-example v0.1.0 (/private/var/folders/kb/tw_lp_xd2_bbv0hqk4m0bvt80000gn/T/tmp.jfRdy7s1HQ/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.58s
     Running `target/debug/killed-9-example`
Killed: 9
$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/debug/killed-9-example`
Killed: 9
$ cargo clean
$ cargo run
... it works again ...
```

An even simpler example, with no Rust dependencies:

```console
$ cargo new simpletest
     Created binary (application) `simpletest` package
$ cd simpletest/
$ cat <<'EOF' > shell.nix
{ pkgs ? import <nixpkgs> { } }:
with pkgs;
pkgs.mkShell {
  nativeBuildInputs = [
    cargo
    cargo-edit
    libiconv
    openssl
    pkg-config
    rust-analyzer
    rustc
    rustfmt
    darwin.apple_sdk.frameworks.Security
  ];
}
EOF
$ nix-shell
$ cargo run
   Compiling simpletest v0.1.0 (/private/var/folders/kb/tw_lp_xd2_bbv0hqk4m0bvt80000gn/T/tmp.n5qubIFyw2/simpletest)
    Finished dev [unoptimized + debuginfo] target(s) in 0.24s
     Running `target/debug/simpletest`
Hello, world!
$ nvim -u NONE src/main.rs
... trivial change, "hello world" -> "hello worlds"
$ cargo run
   Compiling simpletest v0.1.0 (/private/var/folders/kb/tw_lp_xd2_bbv0hqk4m0bvt80000gn/T/tmp.n5qubIFyw2/simpletest)
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
     Running `target/debug/simpletest`
Killed: 9
$ echo $?
137
```

### Expected behavior

`cargo run` works without `Killed: 9`

### Screenshots

I think the above examples should suffice

### Notify maintainers

@Mic92 @LnL7 @zowoq

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 21.3.0, macOS 12.2.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.5pre20211007_844dd90`
 - channels(n8henrie): `""`
 - channels(root): `"darwin, nixpkgs-22.05pre358510.d34f7085a90"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

**Comment by Mic92** [comment]
Can you replace `rustc` with `rustup` and still reproduce the issue (when the same rust version is used)?

**Comment by n8henrie** [comment]
I thought it might be related to my editor interacting with rust-analyzer somehow, but it looks like that's not the case:

```console
$ nix-shell --pure
[nix-shell:~/git/killed-9-example]$ cat src/main.rs
fn main() {
    println!("Hello, world!");
}
[nix-shell:~/git/killed-9-example]$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
     Running `target/debug/killed-9-example`
Hello, world!

[nix-shell:~/git/killed-9-example]$ sed -i 's/world/you/' src/main.rs

[nix-shell:~/git/killed-9-example]$ cat src/main.rs
fn main() {
    println!("Hello, you!");
}

[nix-shell:~/git/killed-9-example]$ cargo run
   Compiling killed-9-example v0.1.0 (/Users/n8henrie/git/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
     Running `target/debug/killed-9-example`
Killed: 9
```

> Can you replace rustc with rustup and still reproduce the issue

Yes:

Before:
```console
[nix-shell:~/git/killed-9-example]$ rustc --version
rustc 1.58.1
```

After:
```console
$ sed -i 's/rustc/rustup/' shell.nix
$ echo '1.58.1' > rust-toolchain
$ nix-shell --pure
[nix-shell:~/git/killed-9-example]$ rustc --version
rustc 1.58.1 (db9d1b20b 2022-01-20)

[nix-shell:~/git/killed-9-example]$ cargo run
   Compiling killed-9-example v0.1.0 (/Users/n8henrie/git/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.85s
     Running `target/debug/killed-9-example`
Hello, you!

[nix-shell:~/git/killed-9-example]$ sed -i 's/you/world/' src/main.rs

[nix-shell:~/git/killed-9-example]$ cat src/main.rs
fn main() {
    println!("Hello, world!");
}

[nix-shell:~/git/killed-9-example]$ cargo run
   Compiling killed-9-example v0.1.0 (/Users/n8henrie/git/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.21s
     Running `target/debug/killed-9-example`
Killed: 9
```

**Comment by Mic92** [comment]
Ok. Could this not be a bug of the library rather than rustc/cargo? You could try to attach lldb and report the backtrace to the crate author.

**Comment by n8henrie** [comment]
Also, I don't see this behavior outside of `nix-shell`:

```console
$ cargo new foo
     Created binary (application) `foo` package
$ cd foo/
$ cargo run
   Compiling foo v0.1.0 (foo)
    Finished dev [unoptimized + debuginfo] target(s) in 0.74s
     Running `target/debug/foo`
Hello, world!
$ sed -i 's/world/you/' src/main.rs
$ cargo run
   Compiling foo v0.1.0 (foo)
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/foo`
Hello, you!
```

**Comment by sternenseemann** [comment]
`log(1)` could help you determine what exactly gets killed and why.

**Comment by n8henrie** [comment]
Thank you for the suggestion. I'll take a look and report back soon.

**Comment by n8henrie** [comment]
The following lines look suspicious, but this is above my level of expertise. I don't understand why changing a single line in the source code would produce this error in nix-shell but not in MacOS.

```
kernel: CODE SIGNING: cs_invalid_page(0x104ccc000): p=23225[killed-9-example] final status 0x23020200, denying page sending SIGKILL
kernel: CODE SIGNING: process 23225[killed-9-example]: rejecting invalid page at address 0x104ccc000 from offset 0x38000 in file "/
```

Full output:

```console
$ log show --last 10m --debug --predicate 'eventMessage contains "killed-9-example"'
Timestamp                       Thread     Type        Activity             PID    TTL  
2022-03-19 08:45:39.172822-0600 0x1a566    Default     0x0                  0      0    kernel: (AppleMobileFileIntegrity) AMFI: '/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example' has no CMS blob?
2022-03-19 08:45:39.172829-0600 0x1a566    Default     0x0                  0      0    kernel: (AppleMobileFileIntegrity) AMFI: '/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example': Unrecoverable CT signature issue, bailing out.
2022-03-19 08:45:39.173474-0600 0x1a568    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) allowed via dev tool environment(/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:39.173979-0600 0x1a568    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libiconv.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:39.174004-0600 0x1a568    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libiconv.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:39.174091-0600 0x1a568    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libiconv-nocharset.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:39.174114-0600 0x1a568    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libiconv-nocharset.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:39.174186-0600 0x1a568    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libcharset.1.0.0.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:39.174211-0600 0x1a568    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libcharset.1.0.0.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:54.948309-0600 0x1a86f    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) allowed via dev tool environment(/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:54.948758-0600 0x1a86f    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libiconv.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:54.948782-0600 0x1a86f    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libiconv.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:54.948861-0600 0x1a86f    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libiconv-nocharset.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:54.948879-0600 0x1a86f    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libiconv-nocharset.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:54.948946-0600 0x1a86f    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libcharset.1.0.0.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:54.948964-0600 0x1a86f    Debug       0x0                  0      0    kernel: (AppleSystemPolicy) Library (/nix/store/sv54j4kykmp0v3izfnd8ycl8kli4l2ij-libiconv-50/lib/libcharset.1.0.0.dylib) allowed in Process (/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example) via dev tool environment (/System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal)
2022-03-19 08:45:54.949439-0600 0x1a86f    Default     0x0                  0      0    kernel: CODE SIGNING: cs_invalid_page(0x104ccc000): p=23225[killed-9-example] final status 0x23020200, denying page sending SIGKILL
2022-03-19 08:45:54.949452-0600 0x1a86f    Default     0x0                  0      0    kernel: CODE SIGNING: process 23225[killed-9-example]: rejecting invalid page at address 0x104ccc000 from offset 0x38000 in file "/Users/n8henrie/git/killed-9-example/target/debug/killed-9-example" (cs_mtime:1646679111.830753257 != mtime:1647701154.916071530) (signed:1 validated:1 tainted:1 nx:0 wpmapped:0 dirty:0 depth:0)
2022-03-19 08:45:54.949513-0600 0x1a86f    Default     0x0                  0      0    kernel: killed-9-example[23225] Corpse allowed 1 of 5
2022-03-19 08:45:55.486546-0600 0xa06f     Default     0x0                  1238   0    ReportCrash: Formulating fatal 309 report for corpse[23225] killed-9-example
2022-03-19 08:45:55.536960-0600 0x1a883    Default     0x0                  23268  0    osanalyticshelper: (OSAnalytics) creating type 309 as /Users/n8henrie/Library/Logs/DiagnosticReports/.killed-9-example-2022-03-19-084555.ips
2022-03-19 08:45:55.650391-0600 0x1a883    Default     0x0                  23268  0    osanalyticshelper: (OSAnalytics) Saved type '309(<private>)' report (1 of max 25) at /Users/n8henrie/Library/Logs/DiagnosticReports/killed-9-example-2022-03-19-084555.ips
2022-03-19 08:45:55.650962-0600 0x1a883    Default     0x0                  23268  0    osanalyticshelper: xpc log creation type 309 result success: /Users/n8henrie/Library/Logs/DiagnosticReports/killed-9-example-2022-03-19-084555.ips
2022-03-19 08:45:55.651118-0600 0xa06f     Default     0x0                  1238   0    ReportCrash: (OSAnalytics) client log create type 309 result success: /Users/n8henrie/Library/Logs/DiagnosticReports/killed-9-example-2022-03-19-084555.ips
2022-03-19 08:45:55.651171-0600 0xa06f     Default     0x0                  1238   0    ReportCrash: no MetricKit for process killed-9-example type 309 bundleId (null)
2022-03-19 08:45:55.651405-0600 0xa06f     Default     0x0                  1238   7    ReportCrash: (CoreAnalytics) [com.apple.CoreAnalytics.stability-event:event-send] Sending event: com.apple.stability.crash {"exceptionCodes":"0x0000000000000032, 0x0000000104ccc000(\n    50,\n    4375494656\n)EXC_BAD_ACCESSSIGKILL (Code Signature Invalid)UNKNOWN_0x32 at 0x0000000104ccc000","incidentID":"1BCF2A6D-3234-4DF2-8E02-524D84272CF3","logwritten":1,"process":"killed-9-example","responsibleApp":"Terminal","terminationReasonExceptionCode":"0x2","terminationReasonNamespace":"CODESIGNING"}
2022-03-19 08:45:55.651663-0600 0x19e4c    Default     0x0                  643    7    analyticsd: [com.apple.CoreAnalytics.stability-event:event-recv] Received event: com.apple.stability.crash {"exceptionCodes":"0x0000000000000032, 0x0000000104ccc000(\n    50,\n    4375494656\n)EXC_BAD_ACCESSSIGKILL (Code Signature Invalid)UNKNOWN_0x32 at 0x0000000104ccc000","incidentID":"1BCF2A6D-3234-4DF2-8E02-524D84272CF3","logwritten":1,"process":"killed-9-example","responsibleApp":"Terminal","terminationReasonExceptionCode":"0x2","terminationReasonNamespace":"CODESIGNING"}
2022-03-19 08:45:55.656449-0600 0x19e4c    Default     0x0                  643    7    analyticsd: [com.apple.CoreAnalytics.stability-event:event-aggregated] Aggregated. Transform: StabilityCrashNumerator3WithBundleVersion Dirty: 1 Event: com.apple.stability.crash {"exceptionCodes":"0x0000000000000032, 0x0000000104ccc000(\n    50,\n    4375494656\n)EXC_BAD_ACCESSSIGKILL (Code Signature Invalid)UNKNOWN_0x32 at 0x0000000104ccc000","incidentID":"1BCF2A6D-3234-4DF2-8E02-524D84272CF3","logwritten":1,"process":"killed-9-example","responsibleApp":"Terminal","terminationReasonExceptionCode":"0x2","terminationReasonNamespace":"CODESIGNING","timestamp":1647701155628861}
2022-03-19 08:45:55.656556-0600 0x19e4c    Default     0x0                  643    7    analyticsd: [com.apple.CoreAnalytics.stability-event:event-aggregated] Aggregated. Transform: StabilityCrashNumerator3WithIncidentID Dirty: 1 Event: com.apple.stability.crash {"exceptionCodes":"0x0000000000000032, 0x0000000104ccc000(\n    50,\n    4375494656\n)EXC_BAD_ACCESSSIGKILL (Code Signature Invalid)UNKNOWN_0x32 at 0x0000000104ccc000","incidentID":"1BCF2A6D-3234-4DF2-8E02-524D84272CF3","logwritten":1,"process":"killed-9-example","responsibleApp":"Terminal","terminationReasonExceptionCode":"0x2","terminationReasonNamespace":"CODESIGNING","timestamp":1647701155628861}
2022-03-19 08:45:55.656654-0600 0x19e4c    Default     0x0                  643    7    analyticsd: [com.apple.CoreAnalytics.stability-event:event-aggregated] Aggregated. Transform: StabilityCrashNumerator3 Dirty: 1 Event: com.apple.stability.crash {"exceptionCodes":"0x0000000000000032, 0x0000000104ccc000(\n    50,\n    4375494656\n)EXC_BAD_ACCESSSIGKILL (Code Signature Invalid)UNKNOWN_0x32 at 0x0000000104ccc000","incidentID":"1BCF2A6D-3234-4DF2-8E02-524D84272CF3","logwritten":1,"process":"killed-9-example","responsibleApp":"Terminal","terminationReasonExceptionCode":"0x2","terminationReasonNamespace":"CODESIGNING","timestamp":1647701155628861}
```

**Comment by a-h** [comment]
I have an M1 Mac, so I tried out your example repo, and couldn't reproduce the problem. 

https://asciinema.org/a/gHz8aFVsGOGaG8xbB9txIHIB4

**Comment by n8henrie** [comment]
@a-h thanks for chiming in! That's good to know, thanks for trying to reproduce.

**Comment by n8henrie** [comment]
@a-h Also -- multi-user install?

```console
$ nix --version
nix (Nix) 2.5pre20211007_844dd90
```

**Comment by a-h** [comment]
I was on `nix (Nix) 2.3.16` (multi-user) but I decided to update to test with later versions.

I removed Nix from my Mac, and reinstalled which brought in `nix (Nix) 2.7.0` and also brought in a later version of rust (rustc 1.58.1) presumably due to updating channels, since once I started using it, I could run your example without first dropping the rust code in the toml file down to 2018.

However, still no problems for me. I tried it at least 10 times. I don't think it would make any difference about exactly which Mac I'm using, but I'm on an M1 Max MacBook Pro.

I don't use Nix in a great way. I'm usually the only person on a team that's using it, and I mainly just use it configure my system. My entire config is here: https://github.com/a-h/dotfiles in case there's anything useful in there for you.

You don't have a virus scanner running do you? Killing a program randomly like that sounds like the sort of thing it might do.

**Comment by thefloweringash** [comment]
I can reproduce the problem on recent `master` (f5054121cb287317c4ca8c409ef9c68f36658013), but not in a way that makes it easy to debug.

```
[nix-shell:~/src/killed-9-example]$ i=0; while /run/current-system/sw/bin/git checkout src/main.rs && sed -i "s/world/$((i++))/" src/main.rs && cargo run; do :; done

[... snip runs 0 to 467 ...]

Updated 1 path from the index
   Compiling killed-9-example v0.1.0 (/Users/lorne/src/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.17s
     Running `target/debug/killed-9-example`
Hello, 468!
Updated 1 path from the index
   Compiling killed-9-example v0.1.0 (/Users/lorne/src/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
     Running `target/debug/killed-9-example`
Killed: 9
```

But this only seems to be a transient error, and the resulting binary is valid if I try to run it again:
```
[nix-shell:~/src/killed-9-example]$ target/debug/killed-9-example
Hello, 469!
```

<details>
<summary>Versions</summary>

```
❯ nix --version
nix (Nix) 2.7.0

❯ sw_vers
ProductName:	macOS
ProductVersion:	12.3
BuildVersion:	21E230

❯ nix-daemon --version
nix-daemon (Nix) 2.7.0
```

**Comment by a-h** [comment]
Ah, I wasn't persistent enough. I did think about scripting it, but didn't follow through.

Using your little script, I was also able to reproduce the error after 843 times...

```
Hello, 843!
Updated 1 path from the index
   Compiling killed-9-example v0.1.0 (/Users/adrian/github.com/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.17s
     Running `target/debug/killed-9-example`
Killed: 9
```

**Comment by n8henrie** [comment]
@a-h 

> You don't have a virus scanner running do you?

Nope. Again, thanks for your efforts here.

@jasonrm

> you might try updating your nix channels

No luck. Could using nix-darwin be an issue here?

```console
$ sudo -i nix-channel --list
darwin https://github.com/LnL7/nix-darwin/archive/master.tar.gz
nixpkgs https://nixos.org/channels/nixpkgs-unstable
$ sudo -i nix-channel --update
unpacking channels...
$ nix-shell --pure
[nix-shell:~/git/killed-9-example]$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/killed-9-example`
Killed: 9
```

Tried `cargo clean` and a few reruns, no luck.

@thefloweringash

Odd, it's basically instant for me, and resulting binary is no good if I try to re-execute. Looks like codesigning worked.

```console
$ git checkout src/main.rs
Updated 0 paths from the index
$ nix-shell --pure
[nix-shell:~/git/killed-9-example]$ cargo clean
[nix-shell:~/git/killed-9-example]$ i=0; while cargo run; do sed -i "s/world[0-9]*/world$((i++))/" src/main.rs; done
   Compiling killed-9-example v0.1.0 (/Users/n8henrie/git/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.50s
     Running `target/debug/killed-9-example`
Hello, world!
   Compiling killed-9-example v0.1.0 (/Users/n8henrie/git/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
     Running `target/debug/killed-9-example`
Killed: 9
[nix-shell:~/git/killed-9-example]$ ./target/debug/killed-9-example
Killed: 9
[nix-shell:~/git/killed-9-example]$ codesign
bash: codesign: command not found
[nix-shell:~/git/killed-9-example]$ exit
exit
$ codesign -vv ./target/debug/killed-9-example
./target/debug/killed-9-example: valid on disk
./target/debug/killed-9-example: satisfies its Designated Requirement
```

I don't seem to have `cargo-cbuild`, and my `cargo` and `rustc` are not codesigned. Should they be?

```console
[nix-shell:~/git/killed-9-example]$ command -v cargo-cbuild
[nix-shell:~/git/killed-9-example]$ command -v cargo
/nix/store/sakd69ff7clis6b5w46cxjblgkl7r904-cargo-1.58.1/bin/cargo
[nix-shell:~/git/killed-9-example]$ command -v rustc
/nix/store/sfia0kg9dz6ii1i20jh14ijdpxky98sm-rustup-1.24.3/bin/rustc
[nix-shell:~/git/killed-9-example]$ exit
exit
$ codesign -vv /nix/store/sfia0kg9dz6ii1i20jh14ijdpxky98sm-rustup-1.24.3/bin/rustc
/nix/store/sfia0kg9dz6ii1i20jh14ijdpxky98sm-rustup-1.24.3/bin/rustc: code object is not signed at all
$ codesign -vv /nix/store/sakd69ff7clis6b5w46cxjblgkl7r904-cargo-1.58.1/bin/cargo
/nix/store/sakd69ff7clis6b5w46cxjblgkl7r904-cargo-1.58.1/bin/cargo: code object is not signed at all
```

@a-h

> Because if it happens using a different installation mechanism, it can't be anything to do with Nix.

Well I'll be darned, you're right -- same crash outside of nix. Sorry to waste so much time! I've never seen it outside of nix so I was surprised.

Ends up, I think it's perhaps a problem with `1.58.1-aarch64-apple-darwin`, which I [pinned above](https://github.com/NixOS/nixpkgs/issues/162795#issuecomment-1061017959) but didn't commit to the repo (until just now).

I've now added the toolchain file and an example crasher script to the repo, though I'm also closing the issue as it looks like a rust problem and not nix.

Thanks so much to all for your help and again sorry for the time spent!

**Comment by n8henrie** [comment]
Well, it works *outside* of nix now (`crashme.sh` runs without crashing), likely since it's using 1.59, but still crashes in nix-shell, likely because it's still at 1.58.1 there. ¯\\\_(ツ)_/¯

```console
$ rustc --version
rustc 1.59.0 (9d1b2106e 2022-02-23)
$ nix-shell --pure

[nix-shell:~/git/killed-9-example]$ rustc --version
rustc 1.58.1

[nix-shell:~/git/killed-9-example]$ cat rust-toolchain.toml
[toolchain]
# channel = "1.58.1"
channel = "stable"

[nix-shell:~/git/killed-9-example]$ ./crashme.sh
Updated 1 path from the index
   Compiling killed-9-example v0.1.0 (/Users/n8henrie/git/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.32s
     Running `target/debug/killed-9-example`
Hello, world!
   Compiling killed-9-example v0.1.0 (/Users/n8henrie/git/killed-9-example)
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
     Running `target/debug/killed-9-example`
./crashme.sh: line 10: 75979 Killed: 9               cargo run
```

Some seemingly related issues:

- https://github.com/rust-lang/cargo/issues/10060
- https://github.com/rust-lang/rust/issues/91541

I suppose I should still close, with the only pending issue being to wait for nixpkgs-unstable to update to 1.59.0?

**Comment by sternenseemann** [comment]
#165406 has the rustc update iirc.

**Comment by n8henrie** [comment]
I'm wondering if this may have been related to using <https://github.com/michaeleisel/zld/issues/110> the whole time; today I started getting the same crashes *outside* of nix, using rustc 1.60.0.

I've updated my example repo to a flake: <https://github.com/n8henrie/killed-9-example>

Interestingly, I can run `nix develop -i` in that repo and reproduce the crash -- but when I remove the `zld` relevant config from `~/.cargo/config`, the crash also resolves in the nix environment (I guess not a nix shell anymore? what's the terminology here?).

Why is that? I had thought that `nix develop -i` should be *reasonably* close to `nix-shell --pure`, and that my `~/.cargo/config` shouldn't even be picked up.

Should I be using something else for highly reproducibility / pure-ness with nix flakes?
